merge.py

```diff
@@ -1,98 +1,320 @@
+        """
+        _print_flush('Cloning repository and configuring.')
+        repo = self.config['repository']
+        repo_url = GITHUB_REPO_URL.format(org='apache', repo=repo)
+        clone_args = ['git', 'clone', '-b', self.config['merge_branch'],
+                      repo_url, tmp_dir]
+        try:
+            check_call(clone_args, cwd=tmp_dir)
+        except:
+            pr.post_error(
+                'Couldn\'t clone from github/{}/{}. Please try again.'.format(
+                    'apache', repo))
+            raise AssertionError('Clone Failed.')
+
+        try:
+            check_call(['git', 'remote', 'add', 'apache',
+                        APACHE_GIT.format(repo=repo)], cwd=tmp_dir)
+            check_call(['git', 'remote', 'rename', 'origin', 'github'],
+                       cwd=tmp_dir)
+            check_call('git config --local --add remote.github.fetch'
+                       ' "+refs/pull/*/head:refs/remotes/github/pr/*"',
+                       shell=True, cwd=tmp_dir)
+            check_call(['git', 'fetch', '--all'], cwd=tmp_dir)
+        except:
+            pr.post_error('Problem configuring repository. Please try again.')
+            raise AssertionError('Failed to set up.')
+        _print_flush('Initial configuration complete.')
+
+    def checkout(self, pr_name, tmp_dir, pr):
+        """Check out PR from Github.
+
+        Args:
+            pr_name: Name of pull request branch.
+            tmp_dir: Directory in which commands should be run.
+            pr: Pull request to merge.
+        Raises:
+            AssertionError: If command was not successful.
+            EnvironmentError: If PR comment couldn't be posted to Github.
+        """
+        _print_flush('Checking out PR.')
+        pr_num = pr.get_num()
+        ic_args = ['git', 'checkout', '-b', pr_name, 'github/pr/{}'.format(
+            pr_num)]
+        try:
+            check_call(ic_args, cwd=tmp_dir)
+        except:
+            pr.post_error("Couldn't checkout code. Please try again.")
+            raise AssertionError('Failed to check out PR.')
+        _print_flush('Checked out.')
+
+    def rebase(self, branch, tmp_dir, pr):
+        """Rebase PR onto main.
+
+        Args:
+            branch: Path to branch.
+            tmp_dir: Directory in which commands should be run.
+            pr: Pull request to merge.
+        Raises:
+            AssertionError: If command was not successful.
+            EnvironmentError: If PR comment couldn't be posted to Github.
+        """
+        _print_flush('Rebasing onto main.')
+        rebase_args = ['git', 'rebase', branch]
+        try:
+            check_call(rebase_args, cwd=tmp_dir)
+        except:
+            pr.post_error('Automatic rebase was not successful. Please rebase '
+                          'against main and try again.')
+            raise AssertionError('Failed to rebase.')
+        _print_flush('Rebased successfully.')
```

Comment:
Continuing the thread from above, if the "execute this sequence of commands" method also took a description, then it is easy to do the:

1. Print that we're about to do this.
2. Run the sequence of commands. Report a failure if it happens
3. Print that we did this successfully.